### ISSUE TITLE
Why subbn works?

### ISSUE THREAD

**Issue by SuX97** [facebookresearch/SlowFast#276]
Hi, thank you for your great work!

I am trying to understand how multigrid works, especially the subbn module.

The `num_features` of split_bn will be different at each long cycle, to aggregate the running_mean and var, what's been done in the code is to fold the vector of mean at channel-dimension and average it, what's the philosophy behind it (why does the mean of the first 8 channels' average can represent the 1st channel)? why not just don't change the num_channels?

And in resuming from the checkpoint, you simply concat the mean and std for `num_splits` times. Again I cannot get the point of that. It seems to me that these operations will destroy all the representations learned so far, and the model has to slowly "recover" and "adapt to" this change, and maybe that why finetune stage is extremely important. Again why not just don't change the num_channels?

Besides, as for as I am concerned, the named_buffers(especially num_batch_tracked) relies on saving/loading checkpoints. in Sthv2 config, the CHECKPOINT_PERIOD is 2, will the named_buffer of bns loss track if the longcycle switching happen to be 2 epochs later than the last saving?

**Comment by OValery16** [comment]
i m also interested

**Comment by chaoyuaw** [comment]
Hi @SuX97 and @OValery16 , Thanks for the questions and thanks for trying multigrid! 

1. What sub-bn does is to have multiple copies of BNs (sharing weights), each of which runs on a "group" of examples within a GPU. In our code, we however don't explicitly split the batch and construct multiple BNs, because that'll be slow. What we do is to first "reshape" to fold examples to the channel dimension (so that group 1 corresponds to channels [0, C-1], group 2 corresponds to channels [C, 2C - 1], and so on). We then apply BN, and reshape it back (Please see https://github.com/facebookresearch/SlowFast/blob/master/slowfast/models/batchnorm_helper.py#L101).
In this way, BN is correctly applied to each group of examples. Also note that the weight and bias are applied after reshaping it back, so weight/bias are shared across groups.  The only side effect is that this creates multiple copies of "running statistics". We thus need to aggregate them. Aggregation is straightforward, because channels [0, C-1], channels [C, 2C-1], etc. are semantically the same thing.

2. When resuming from checkpoints, we concatenate stats, because channels [0, C-1], channels [C, 2C-1], etc. essentially correspond to multiple copies of the same BN layer.

3. When multigrid is used, checkpoint period has a different rule. Please see
https://github.com/facebookresearch/SlowFast/blob/master/slowfast/utils/checkpoint.py#L84
for more details.

**Comment by SuX97** [comment]
Thank you for your explanation, I still have question about the 2nd answer. 

In `utils/checkpoint.py`, it seems that this concatenation only occur when the channel(namely the num_splits) expand, such as from 1 split to 8 splits, and this occurs only when switching to a new long-cycle(bs_factor from 1 to 8). But within a long-cycle, how does subbn inherit the `running_mean` and `running_var`?
```python
    for key in model_sd:
        if key in checkpoint_sd:
            model_blob_shape = model_sd[key].shape
            c2_blob_shape = checkpoint_sd[key].shape

            if (
                len(model_blob_shape) == 1
                and len(c2_blob_shape) == 1
                and model_blob_shape[0] > c2_blob_shape[0]
                and model_blob_shape[0] % c2_blob_shape[0] == 0
            ):
                before_shape = checkpoint_sd[key].shape
                checkpoint_sd[key] = torch.cat(
                    [checkpoint_sd[key]]
                    * (model_blob_shape[0] // c2_blob_shape[0])
                )
                logger.info(
                    "{} {} -> {}".format(
                        key, before_shape, checkpoint_sd[key].shape
                    )
                )
```

**Comment by chaoyuaw** [comment]
Hi @SuX97 , Thanks for clarifying your questions!

Whenever we switch between long cycle shapes, we save a checkpoint first, reconstruct a new model, and then load the weights just saved to continue training. 

When we save a  checkpoint, we always convert sub-BN layers to normal BN layers before saving, because normal BN layers are what external models/code would expect. (https://github.com/facebookresearch/SlowFast/blob/master/slowfast/utils/checkpoint.py#L123)

Thus, in the case of "shrinking" (say Sub-BN-4 to Sub-BN-2), we store Sub-BN-4 weights with SubBN layers converted to normal BN, (https://github.com/facebookresearch/SlowFast/blob/master/slowfast/utils/checkpoint.py#L123). 
And when loading the weights, we convert the stored normal BN weights into SubBN-2
https://github.com/facebookresearch/SlowFast/blob/master/slowfast/utils/checkpoint.py#L269
using the concatenation method we discussed earlier. 

Does it make sense?

**Comment by SuX97** [comment]
Hi @chaoyuaw, thank you so much for your explanation, it makes sense to me now!
